sopt/src/main/scala/dagr/sopt/cmdline/CommandLineParser.scala

```diff
@@ -0,0 +1,527 @@
+            printExtraUsage(clazz=Some(clazz))
+            print(clpParser.usage(withPreamble=false))
+            None
+          case ParseResult.Version => // Case 7
+            print(clpParser.version)
+            None
+          case ParseResult.Success => // Case 8
+            val clp: Clp = clpParser.instance.get
+            try {
+              clpBlock(clp)
+              Some(clp)
+            }
+            catch {
+              case ex: ValidationException =>
+                printExtraUsage(clazz=Some(clazz))
+                print(clpParser.usage(withPreamble=false))
+                ex.messages.foreach(msg => print(wrapError(msg)))
+                None
+            }
+        }
+    }
+  }
+
+  /** Method for parsing a main classe's command line arguments as well as a specific command line program's arguments.
+    * Typically, the arguments are the format `[main arguments] [Clp Name] [clp arguments]`.  This in constrat to [[parseClp]],
+    * which omits the `[main arguments]` section.
+    *
+    *   (1) Searches for `--` or a recognized [[Clp]] in the list of arguments to partition the main and clp arguments. The
+    *       [[Clp]]s are found by searching for all classes that extend [[Clp]].
+    *   (2) Parses the main arguments and creates an instance of main, and returns None if any error occurred.
+    *   (3) Executes the `mainBlock` of code after the main instance has been created.
+    *   (3) Parses the clp arguments and creates an instance of clp, and returns None if any error occurred.
+    *   (3) Executes the `clpBlock` of code after the clp instance has been created.
+    *
+    * It creates [[CommandLineProgramParser]]s to parse the main class and clp arguments respectively.
+    *
+    * @param args              the command line arguments to parse.
+    * @param packageList       the list of packages to search for classes that extend [[Clp]].
+    * @param omitSubClassesOf  individual classes to omit from including on the command line.
+    * @param includeHidden     true to include classes whose [[clp]] annotation's hidden values is set to false.
+    * @param mainBlock         a block of code to execute after a [[MainClass]] (main) instance has been created successfully.
+    * @param clpBlock          a block of code to execute after a [[clp]] instance has been created successfully.
+    * @return                  a [[clp]] instance initialized with the command line arguments, or [[None]] if
+    *                          unsuccessful.
+    */
+  def parseMainAndClp[MainClass](args: Array[String],
+                                 packageList: List[String],
+                                 omitSubClassesOf: Iterable[Class[_]] = Nil,
+                                 includeHidden: Boolean = false,
+                                 mainBlock: MainClass => Unit.type = (t: MainClass) => Unit,
+                                 clpBlock: MainClass => Clp => Unit.type = (t: MainClass) => (c: Clp) => Unit)
```

Comment:
The way `clpBlock` is specified will require invokers of this method to write code like:
```scala
... clpBlock = (main) => (clp) => { someCode }
```
instead of the following which seems more intuitive:
```scala
... clpBlock = (main, clp) => { someCode }
```